### PR TITLE
[CLOUDGA-13712][CLOUDGA-13713][CLOUDGA-13714][CLOUDGA-13715][CLOUDGA-13716][CLOUDGA-13717] Add CLI role API management

### PR DESCRIPTION
**Pull request by harshitha-yb**
Adding CLI support for list roles, describe role, create role, update role, delete role and managing the custom role feature flag while using these APIs.

1. List Roles 
Screenshot 2023-05-17 at 12 00 37 AM

2. Describe role
Screenshot 2023-05-17 at 12 32 05 AM
Screenshot 2023-05-17 at 12 32 27 AM
Screenshot 2023-05-17 at 12 37 44 AM

3. Create and update role
Screenshot 2023-05-17 at 12 54 41 AM

4. Delete role
Screenshot 2023-05-17 at 1 06 07 AM

### CODE REVIEW

File: internal/client/client.go [hunk 1]
```diff
@@ -775,6 +775,10 @@ func (a *AuthApiClient) ListTasks() ybmclient.ApiListTasksRequest {
 	return a.ApiClient.TaskApi.ListTasks(a.ctx, a.AccountID)
 }
 
+func (a *AuthApiClient) ListRbacRoles() ybmclient.ApiListRbacRolesRequest {
+	return a.ApiClient.RoleApi.ListRbacRoles(a.ctx, a.AccountID).RoleTypes("ALL").Limit(100)
+}
```
Comment (harshitha-yb) [inline internal/client/client.go:780]: Hi
So we have a feature flag for custom roles that affects the behaviour of this API
If the feature flag is off and we call this API with the "ALL" role types filter- we get a "404: Requested API not found"
In such a case I want to catch this NFE and try to call this API with the "SYSTEM" filter to try to list only the system roles
Whats the best way to go about this? Thanks in advance :)

Comment (posriniv) [reply internal/client/client.go:780]: Maybe catch the error here and retry the request? https://github.com/yugabyte/ybm-cli/pull/135/files#diff-18363742b4f67cb68058e1b926b87dfdbeb6287fc1cd5bbfbfb061461b0d6455R58-R61

Comment (harshitha-yb) [reply internal/client/client.go:780]: Please let me know if the implementation looks okay, thanks!

File: cmd/role/role.go [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+			if strings.TrimSpace(ybmAuthClient.GetApiErrorDetails(err)) == strings.TrimSpace(util.GetCustomRoleFeatureFlagDisabledError()) {
+				systemRoleListRequest := authApi.ListSystemRbacRolesWithPermissions()
+				systemRoleListRequest = systemRoleListRequest.DisplayName(roleName)
+				respTwo, rTwo, errTwo := systemRoleListRequest.Execute()
+
+				if errTwo != nil {
+					logrus.Debugf("Full HTTP response: %v", rTwo)
+					logrus.Fatalf(ybmAuthClient.GetApiErrorDetails(errTwo))
+				} else {
+					resp = respTwo
```
Comment (posriniv) [inline cmd/role/role.go:119]: nit: same as above, maybe better names?

**Review by posriniv — COMMENTED** [review]
Looks okay to me apart from the minor comments. Would request @cdavid also to take a look.

File: cmd/role/role.go [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+		roleName, _ := cmd.Flags().GetString("role-name")
+		if roleName != "" {
+			roleListRequest = roleListRequest.DisplayName(roleName)
+		}
+
+		resp, r, err := roleListRequest.Execute()
+
+		if err != nil {
+			if strings.TrimSpace(ybmAuthClient.GetApiErrorDetails(err)) == strings.TrimSpace(util.GetCustomRoleFeatureFlagDisabledError()) {
+				systemRoleListRequest := authApi.ListSystemRbacRoles()
```
[the remark this answers is not in the document]
Comment (harshitha-yb) [reply cmd/role/role.go:62]: GetRoleByName will not work without explicitly having the role name filter
I was gonna make a common function for the repeated case in list and describe but describe needs the permissions included so thats a different call

**Comment by posriniv** [comment]
Seems like there is a problem with the gingko CLI version
```
Ginkgo detected a version mismatch between the Ginkgo CLI and the version of Ginkgo imported by your packages:
  Ginkgo CLI Version:
    2.9.2
  Mismatched package versions found:
    2.9.5 used by ybm-cli, cmd, read-replica, client, formatter, tabwriter, templates, releases

```

**Comment by posriniv** [comment]
@harshitha-yb The remote ginkgo version got updated and it is getting fetched while running the UTs, however, the go.sum file is not getting updated. We need to have `go mod tidy` in `make test`.  That fixed the problem. cc: @cdavid @shinji62

File: cmd/role/role.go [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+		}
+
+		formatter.SingleRoleWrite(rolesCtx, resp.GetData())
+	},
+}
+
+func CreatePermissionsMap(resourceType string, operationGroup string, permissionsMap map[string][]string) {
+	if ops, ok := permissionsMap[resourceType]; ok {
+		ops = append(ops, operationGroup)
+		permissionsMap[resourceType] = ops
```
Comment (cdavid) [inline cmd/role/role.go:225]: What if you add the same permission multiple times for the same resource? What does the API do in that case?

File: cmd/role/role.go [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+		authApi, err := ybmAuthClient.NewAuthApiClient()
+		if err != nil {
+			logrus.Fatalf("could not initiate api client: %s", err.Error())
+		}
+		authApi.GetInfo("", "")
+
+		roleName, _ := cmd.Flags().GetString("role-name")
+		roleId, err := authApi.GetRoleIdByName(roleName)
+		if err != nil {
+			logrus.Fatal(err)
```
Comment (cdavid) [inline cmd/role/role.go:245]: I would add an extra: `Unable to find role with name {}`

File: cmd/role/role.go [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+		}
+
+		permissionsMap := map[string][]string{}
+
+		permissionsList, _ := cmd.Flags().GetStringArray("permissions")
+		for _, permissionsString := range permissionsList {
```
Comment (cdavid) [inline cmd/role/role.go:251]: This section seems to be common between create/update of role? Any chance we can reuse the code?

File: internal/client/client.go [hunk 7]
```diff
@@ -775,6 +775,96 @@ func (a *AuthApiClient) ListTasks() ybmclient.ApiListTasksRequest {
 	return a.ApiClient.TaskApi.ListTasks(a.ctx, a.AccountID)
 }
 
+func (a *AuthApiClient) ListRbacRoles() ybmclient.ApiListRbacRolesRequest {
```
Comment (cdavid) [inline internal/client/client.go:778]: Nitpick: `ListAllRbacRoles()` ?

**Review by cdavid — APPROVED** [review]
Please wait for @shreyas-agnihotri to sign off on the doc. @ddhodge FYI, we will have some doc changes.

File: cmd/role/role.go (hunk 5 above)
Comment (harshitha-yb) [reply cmd/role/role.go:245]: As a logrus.Debug? 
Or print it for the user?

File: cmd/role/role.go (hunk 4 above)
Comment (harshitha-yb) [reply cmd/role/role.go:225]: It just ignores them
Only takes it into account once

**Pull request merged**